src/pydsl/scf.py

```diff
@@ -2,12 +2,299 @@
+    """
+    Base class for constructing SCF ops (IfOp and ForOp).
+    Subclasses implement static methods:
+    - build_op(node, op_type, op_args, dummy)
+    - build_regions(visitor, node, op)
+    - excluded_vars(node) -> tuple
+    """
+
+    def __new__(cls, visitor, node):
+        scope = visitor.scope_stack
+        current_scope = scope.stack[-1]
+        bound_prior = current_scope.f_locals
+
+        # 1. Preprocess node
+        cls.preprocess_node(visitor, node)
+
+        # 2. Build dummy op
+        dummy_op = cls.build_op(node, op_type=[], op_args=[], dummy=True)
+
+        # 3. Build region bindings
+        bound_dicts = cls.build_regions(visitor, node, dummy_op)
+
+        # 4. Analyze yields
+        yield_vars_cand: list[str] = [
+            var
+            for d in bound_dicts
+            for var, val in d.items()
+            if val != bound_prior.get(var)
+            and var not in cls.excluded_vars(node)
+        ]
+
+        op_type: list[type] = []
+        yield_vars = []
+        yield_vals_per_region: list[list[any]] = [[] for _ in bound_dicts]
+
+        for var in yield_vars_cand:
+            types = []
+            if var in bound_prior:
+                types.append(type(bound_prior[var]))
+            for d in bound_dicts:
+                if var in d:
+                    types.append(type(d[var]))
+
+            # TODO be more lenient and assign poison value after op
+            if len(set(types)) > 1:
+                raise TypeError(
+                    f"variable '{var}' has mismatched types: {types}"
+                )
+
+            yield_vars.append(var)
+
+            rst_type = types[0] if types else object
+            op_type.append(rst_type)
+            for i, d in enumerate(bound_dicts):
+                val = d.get(var) or bound_prior[var]
+                yield_vals_per_region[i].append(val)
+
+        # 5. Build final op
+        new_op = cls.build_op(
+            node,
+            op_type,
+            [bound_prior.get(v) for v in yield_vars],
+            dummy=False,
+        )
+
+        # 6. Move blocks
+        for dummy_region, new_region, vals in zip(
+            dummy_op.regions, new_op.regions, yield_vals_per_region
+        ):
+            move_block(
+                dummy_region.blocks[0],
+                new_region.blocks[0],
+                [lower_single(v) for v in vals],
+            )
+
+        # 7. Replace operands with block arguments if applicable
+        cls._replace_operands_with_block_args(
+            node, new_op, yield_vars, bound_prior
+        )
+
+        # 8. Assign results back
+        for var, typ, val in zip(
+            yield_vars, op_type, new_op.results, strict=True
+        ):
+            scope.assign_name(var, typ(val))
+
+        dummy_op.erase()
+        return new_op
+
+    @staticmethod
+    def preprocess_node(visitor, node):
+        """Optional preprocessing step; subclasses may override"""
+        pass
+
+    @staticmethod
```

Comment:
if you give a function the @staticmethod decorator, the first argument should be `cls` (or class). You'll need to add this argument to all the staticmethods